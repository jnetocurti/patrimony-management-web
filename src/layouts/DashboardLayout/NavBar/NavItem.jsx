import React, { useState, Fragment } from 'react';
import { NavLink as RouterLink } from 'react-router-dom';

import {
  Button,
  Collapse,
  List,
  ListItem,
  makeStyles
} from '@material-ui/core';
import { ExpandLess, ExpandMore } from '@material-ui/icons';

import clsx from 'clsx';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
  item: {
    display: 'flex',
    paddingTop: theme.spacing(0),
    paddingBottom: theme.spacing(0)
  },
  nested: {
    display: 'flex',
    paddingTop: theme.spacing(0),
    paddingBottom: theme.spacing(0),
    paddingLeft: theme.spacing(2)
  },
  button: {
    color: theme.palette.text.secondary,
    fontWeight: theme.typography.fontWeightMedium,
    justifyContent: 'flex-start',
    letterSpacing: theme.spacing(0),
    padding: '8px 6px',
    textTransform: 'none',
    width: '100%'
  },
  icon: {
    marginRight: theme.spacing(1)
  },
  title: {
    marginRight: 'auto'
  },
  active: {
    color: theme.palette.primary.main,
    '& $title': {
      fontWeight: theme.typography.fontWeightMedium
    },
    '& $icon': {
      color: theme.palette.primary.main
    }
  }
}));

const NavItem = ({ config }) => {
  const classes = useStyles();
  const [open, setOpen] = useState(config.open || false);

  const handleClick = () => setOpen(!open);

  const createItem = (options) => {
    const Icon = options.icon;
    const itemClass = options.isParent ? classes.item : classes.nested;
    return (
      <>
        <ListItem
          disableGutters
          className={clsx(itemClass)}
          onClick={options.handleClick}
        >
          <Button
            to={options.href}
            component={RouterLink}
            className={classes.button}
            activeClassName={classes.active}
          >
            {Icon && <Icon className={classes.icon} size="20" />}
            <span className={classes.title}>{options.title}</span>
          </Button>
          {options.isParent
            && options.subItems
            && (open ? <ExpandLess /> : <ExpandMore />)}
        </ListItem>
        {options.subItems && (
          <Collapse in={open} timeout="auto" unmountOnExit>
            <List component="div" disablePadding>
              {options.subItems.map((item) => (
                <Fragment key={item.title}>
                  {createItem({
                    isParent: false,
                    href: item.href,
                    icon: item.icon,
                    title: item.title
                  })}
                </Fragment>
              ))}
            </List>
          </Collapse>
        )}
      </>
    );
  };

  return (
    <>
      {createItem({
        handleClick,
        href: config.href,
        icon: config.icon,
        title: config.title,
        subItems: config.subItems,
        isParent: !!config.subItems
      })}
    </>
  );
};

NavItem.propTypes = {
  config: PropTypes.shape({
    open: PropTypes.bool,
    href: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    icon: PropTypes.elementType.isRequired,
    subItems: PropTypes.arrayOf(
      PropTypes.shape({
        href: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        icon: PropTypes.elementType.isRequired
      })
    )
  })
};

export default NavItem;
