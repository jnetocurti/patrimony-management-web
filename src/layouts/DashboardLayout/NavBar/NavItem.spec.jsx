import React from 'react';

import { ListItem } from '@material-ui/core';
import { Apartment as ApartmentIcon, ListOutlined } from '@material-ui/icons';

import { shallow } from 'enzyme';

import NavItem from './NavItem';

describe('src/layouts/DashboardLayout/NavBar/NavItem', () => {
  const config = {
    icon: ApartmentIcon,
    title: 'Fundos imobiliÃ¡rios',
    href: '/app/real-estate-funds',
    subItems: [
      {
        icon: ListOutlined,
        title: 'Ranking',
        href: '/app/real-estate-funds/ranking'
      }
    ]
  };
  it('should render correctly', () => {
    const component = shallow(<NavItem {...{ config }} />);
    expect(component).toMatchSnapshot();
  });

  it('should render correctly with open property', () => {
    const component = shallow(<NavItem {...{ config }} />);
    component.find(ListItem).at(0).simulate('click');

    expect(component).toMatchSnapshot();
  });
});
