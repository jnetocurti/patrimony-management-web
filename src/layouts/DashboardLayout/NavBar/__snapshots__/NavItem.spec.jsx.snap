// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/layouts/DashboardLayout/NavBar/NavItem should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))
    className="makeStyles-item-1"
    disableGutters={true}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Button))
      activeClassName="makeStyles-active-6"
      className="makeStyles-button-3"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "NavLink",
          "propTypes": Object {
            "activeClassName": [Function],
            "activeStyle": [Function],
            "aria-current": [Function],
            "className": [Function],
            "onClick": [Function],
            "replace": [Function],
            "state": [Function],
            "style": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/app/real-estate-funds"
    >
      <Memo(ForwardRef(ApartmentIcon))
        className="makeStyles-icon-4"
        size="20"
      />
      <span
        className="makeStyles-title-5"
      >
        Fundos imobiliários
      </span>
    </WithStyles(ForwardRef(Button))>
    <Memo(ForwardRef(ExpandMoreIcon)) />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Collapse))
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(List))
      component="div"
      disablePadding={true}
    >
      <WithStyles(ForwardRef(ListItem))
        className="makeStyles-nested-2"
        disableGutters={true}
      >
        <WithStyles(ForwardRef(Button))
          activeClassName="makeStyles-active-6"
          className="makeStyles-button-3"
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "NavLink",
              "propTypes": Object {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "onClick": [Function],
                "replace": [Function],
                "state": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/app/real-estate-funds/ranking"
        >
          <Memo(ForwardRef(ListOutlinedIcon))
            className="makeStyles-icon-4"
            size="20"
          />
          <span
            className="makeStyles-title-5"
          >
            Ranking
          </span>
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(ListItem))>
    </WithStyles(ForwardRef(List))>
  </WithStyles(ForwardRef(Collapse))>
</Fragment>
`;

exports[`src/layouts/DashboardLayout/NavBar/NavItem should render correctly with open property 1`] = `
<Fragment>
  <WithStyles(ForwardRef(ListItem))
    className="makeStyles-item-1"
    disableGutters={true}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(Button))
      activeClassName="makeStyles-active-6"
      className="makeStyles-button-3"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "NavLink",
          "propTypes": Object {
            "activeClassName": [Function],
            "activeStyle": [Function],
            "aria-current": [Function],
            "className": [Function],
            "onClick": [Function],
            "replace": [Function],
            "state": [Function],
            "style": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/app/real-estate-funds"
    >
      <Memo(ForwardRef(ApartmentIcon))
        className="makeStyles-icon-4"
        size="20"
      />
      <span
        className="makeStyles-title-5"
      >
        Fundos imobiliários
      </span>
    </WithStyles(ForwardRef(Button))>
    <Memo(ForwardRef(ExpandLessIcon)) />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(Collapse))
    in={true}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(ForwardRef(List))
      component="div"
      disablePadding={true}
    >
      <WithStyles(ForwardRef(ListItem))
        className="makeStyles-nested-2"
        disableGutters={true}
      >
        <WithStyles(ForwardRef(Button))
          activeClassName="makeStyles-active-6"
          className="makeStyles-button-3"
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "NavLink",
              "propTypes": Object {
                "activeClassName": [Function],
                "activeStyle": [Function],
                "aria-current": [Function],
                "className": [Function],
                "onClick": [Function],
                "replace": [Function],
                "state": [Function],
                "style": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          to="/app/real-estate-funds/ranking"
        >
          <Memo(ForwardRef(ListOutlinedIcon))
            className="makeStyles-icon-4"
            size="20"
          />
          <span
            className="makeStyles-title-5"
          >
            Ranking
          </span>
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(ListItem))>
    </WithStyles(ForwardRef(List))>
  </WithStyles(ForwardRef(Collapse))>
</Fragment>
`;
